contracts/ERC20MultiDelegate.sol

```diff
@@ -0,0 +1,434 @@
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Votes.sol";
+
+/**
+ * @dev A child contract which will be deployed by the ERC20MultiDelegate utility contract
+ * This is a proxy delegator contract to vote given delegatee on behalf of original delegator
+ */
+contract ERC20ProxyDelegator {
+    constructor(ERC20Votes _token, address _delegatee) {
+        _token.approve(msg.sender, type(uint256).max);
+        _token.delegate(_delegatee);
+    }
+}
+
+/**
+ * @dev A utility contract to let delegators to pick multiple delegatee
+ */
+contract ERC20MultiDelegate is ERC1155, Ownable {
+    using Address for address;
+
+    ERC20Votes public token;
+
+    struct DelegateeAmount {
+        address delegatee;
+        uint256 amount;
+    }
+
+    struct ReDepositRecord {
+        uint256[] sourceIds;
+        uint256[] targetIds;
+        uint256[] withdrawnAmounts;
+        uint256[] redepositedAmounts;
+        uint256 sourceIndex;
+        uint256 targetIndex;
+        uint256 remainingSourceAmount;
+        uint256 remainingTargetAmount;
+        uint256 totalRedepositedAmount;
+    }
+
+    /**
+     * @dev Constructor.
+     * @param _token The ERC20 token address
+     * @param _metadata_uri ERC1155 metadata uri
+     */
+    constructor(
+        ERC20Votes _token,
+        string memory _metadata_uri
+    ) ERC1155(_metadata_uri) {
+        token = _token;
+    }
+
+    /**
+     * @dev Deposits and delegates voting power to the multiple delegatees.
+     * @param delegateeAmounts The list of delegatee addresses and corresponding list of ERC20 voting power amount amounts to delegate.
+     */
+    function depositMulti(
+        DelegateeAmount[] calldata delegateeAmounts
+    ) external {
+        uint256 delegateesLength = delegateeAmounts.length;
+
+        require(
+            delegateesLength > 0,
+            "DepositMulti: You should pick at least one delegatee"
+        );
+
+        uint256[] memory ids = new uint256[](delegateesLength);
+        uint256[] memory amounts = new uint256[](delegateesLength);
+
+        for (uint256 index = 0; index < delegateesLength; index++) {
+            address delegatee = delegateeAmounts[index].delegatee;
+            uint256 amount = delegateeAmounts[index].amount;
+
+            createProxyDelegatorAndTransfer(delegatee, amount);
+
+            ids[index] = uint256(uint160(delegatee));
+            amounts[index] = amount;
+        }
+
+        mintBatch(msg.sender, ids, amounts);
+    }
+
+    /**
+     * @dev Re-delegates voting power between delegatees.
+     * @param sources The list of source delegatee addresses and amounts to withdraw.
+     * @param targets The list of target delegatee addresses and amounts to redeposit.
+     * The remaning part if any, will be withdrawn to the user account
+     */
+    function reDeposit(
```

Comment:
I feel like all three methods could be replaced by one, with a list of source and destination tokens, and just have it transfer from or to the ERC20 if there's not enough on one side or the other.